### DISCUSSION TITLE
Minitest assertions for RES

### DISCUSSION THREAD

**Discussion by mostlyobvious** [RailsEventStore/rails_event_store#1867] [Ideas]
Gathering ideas and various implementations of assertions which help when working with RailsEventStore 👇

**Comment by mostlyobvious** [comment]
`arrange`, `act`, `assert_changes` that reads persisted events from given stream:
https://github.com/RailsEventStore/cqrs-es-sample-with-res/blob/a32e18a79b0aaa8e21a9b361aac62df4876b8f49/test/support/test_case.rb#L2-L20
and a quite similar take in https://github.com/arkency/aggregates/blob/af64b361d2c1624e6e9bff86fb8a470d9d0d5c82/test/test_helper.rb#L17-L32

**Comment by mostlyobvious** [comment]
before-after-reject can probably be implemented with https://railseventstore.org/docs/subscribe/#temporary-subscriptions (and subscribe_to_all_events), also being a thread safe version by default

**Comment by mostlyobvious** [comment]
Using a bit older API:
```ruby
    def assert_exact_new_events(event_store, new_events)
      events_so_far = event_store.read_all_streams_forward
      yield
      assert_equal(
        new_events,
        (event_store.read_all_streams_forward - events_so_far).map(&:class)
      )
    end

    def assert_new_events_include(event_store, expected_events)
      events_so_far = event_store.read_all_streams_forward
      yield
      new_events = (event_store.read_all_streams_forward - events_so_far).map(&:class)
      assert(
        (expected_events - new_events).empty?,
        "didn't include all of: #{expected_events} in #{(event_store.read_all_streams_forward - events_so_far).map(&:class)}"
      )
    end
```

**Comment by mostlyobvious** [comment]
```ruby
def assert_items(actual_elements, expected_procs)
  assert_equal(
    expected_procs.length,
    actual_elements.length,
    "number of actual elements does not match expected"
  )
  actual_elements.zip(expected_procs) do |actual, expected|
    expected.call(actual)
  end
end

```

**Comment by mostlyobvious** [comment]
```ruby
def assert_published(event_kind)
  event = event_store.read.of_type(event_kind).last
  assert_not_nil event
  yield event
end

def assert_not_published(event_kind)
  assert_nil event_store.read.of_type(event_kind).last
end
```

**Comment by mostlyobvious** [comment]
https://github.com/RailsEventStore/rails_event_store/issues/659#issuecomment-515233231

^^ that would be more of `assert_dispatched`, there no information about stream when dispatching to handlers

**Comment by tomaszwro** [comment]
Let me throw following idea:

```ruby
assert_has_items(events_of_type(OrderPlaced), [
  -> event {
    assert_equal 1, event.data[:quantity]
  },
])
```

where
- `assert_has_items` is a generic helper for checking if a collection has expected number of items (actually exactly as in above comment https://github.com/RailsEventStore/rails_event_store/issues/659#issuecomment-515236081); if the number is ok, it passes items to respective lambdas, where respective items (events in this case) can have any further assertions
- `events_of_type` just gets you all events of certain type from the event store

**Comment by tomaszwro** [comment]
Additionally above idea can be extended to operate within blocks

```ruby
assert_events_during do
  perform_a_command
end.of_type(OrderPlaced).has_items(
  -> event {
    assert_equal 1, event.data[:quantity]
  }
)

**Comment by mostlyobvious** [comment]
A few more samples from _the wild_:

```ruby
  def assert_equal_event(expected, checked, verify_id: false)
    if verify_id
      assert_equal expected.event_id, checked.event_id
    end
    assert_equal expected.class, checked.class
    assert_equal expected.data, checked.data
  end

  def assert_equal_events(expected, checked, verify_id: false)
    assert_equal expected.size, checked.size
    expected.zip(checked).each do |e, c|
      assert_equal_event(e, c, verify_id: verify_id)
    end
  end

  def assert_equal_unpublished_events(expected, model)
    assert_equal_events(expected, model.unpublished_events)
  end
```

**Comment by mostlyobvious** [comment]
In the meantime I've rebased and merged PR with initial gem. See it at:
https://github.com/RailsEventStore/rails_event_store/tree/master/contrib/minitest-ruby_event_store

**Comment by mostlyobvious** [comment]
Related: https://github.com/palkan/active_event_store/pull/2#issuecomment-916008858
